core/src/main/java/org/elasticsearch/index/translog/Translog.java

```diff
@@ -204,14 +204,30 @@ public Translog(TranslogConfig config) throws IOException {
             }
             foundTranslogs.add(openReader(location.resolve(checkpointTranslogFile), checkpoint));
             Path commitCheckpoint = location.resolve(getCommitCheckpointFileName(checkpoint.generation));
-            Files.copy(location.resolve(CHECKPOINT_FILE_NAME), commitCheckpoint);
-            IOUtils.fsync(commitCheckpoint, false);
-            IOUtils.fsync(commitCheckpoint.getParent(), true);
+            if (Files.exists(commitCheckpoint)) {
+                Checkpoint checkpointFromDisk = Checkpoint.read(commitCheckpoint);
+                if (checkpoint.equals(checkpointFromDisk) == false) {
+                    throw new IllegalStateException("Checkpoint file " + commitCheckpoint.getFileName() + " already exists but has corrupted content expected: " + checkpoint + " but got: " + checkpointFromDisk);
+                }
```

Comment:
well I think we don't even need the second fsync `IOUtils.fsync(commitCheckpoint, false);` since we fsynced the tmp file already so no need to fsync again. ie. if the file exists it's fsynced already. @rmuir WDYT